### PR TITLE
Add HSGP contribution module

### PR DESCRIPTION
**Pull request by juanitorduz**
Closes https://github.com/pyro-ppl/numpyro/issues/1788

**Motivation:** As described in the issue

> I have been using the Hilber Space approximation for Gaussian Processes a lot. The main ingredients are the spectral densities of the kernels and the eigenvalues and eigenvectors of the Laplacian in a box, see https://num.pyro.ai/en/stable/examples/hsgp.html

**Scope of this PR:**
- [x] Basic module and descriiption
- [x] Docstrings
- [x] Tests
- [x] Documentation
- [x] Example Notebook

I would consider this module experimental to get user feedback (as described in the docs). For now, it just supports one-dimensional inputs (my most common use case). In the next iteration, I will add an extension to higher dimensions. I did not want to over-complicate this first PR. 

Happy to get feedback :)

### CODE REVIEW

**Comment by martinjankowiak** [comment]
it'd be great to get feedback from @omarfsosa who wrote the hsgp example in the first place

File: numpyro/contrib/hsgp/approximation.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    :rtype: ArrayImpl
+    """
+    phi = eigenfunctions(x=x, ell=ell, m=m)
+    spd = jnp.sqrt(
+        diag_spectral_density_matern(nu=nu, alpha=alpha, length=length, ell=ell, m=m)
+    )
+    return linear_approximation(phi=phi, spd=spd, m=m, non_centered=non_centered)
+
+
+def hsgp_approximation_periodic_non_centered(
```
Comment (martinjankowiak) [inline numpyro/contrib/hsgp/approximation.py:143]: lots of these method names seem pretty long... do you really need "approximation" everywhere?

Comment (juanitorduz) [reply numpyro/contrib/hsgp/approximation.py:143]: Ups! You are right 🙈. I can remove the “approximation” from the function names.

**Comment by juanitorduz** [comment]
I did not know @omarfsosa was the original author. I give credit to the example in many places, but I can also add his name (or GitHub handle) to the docs as I'm essentially organizing the functions for the example and adding docs and tests 😄

File: numpyro/contrib/hsgp/approximation.py (hunk 1 above)
Comment (juanitorduz) [reply numpyro/contrib/hsgp/approximation.py]: Simplified main function names in https://github.com/pyro-ppl/numpyro/pull/1794/commits/2b1e37d3b34c5d92797ce7f3b08480035cad98ae

**Comment by juanitorduz** [comment]
> I did not know @omarfsosa was the original author. I give credit to the example in many places, but I can also add his name (or GitHub handle) to the docs as I'm essentially organizing the functions from the example and adding docs and tests 😄

Added reference to original author to the module page in https://github.com/pyro-ppl/numpyro/pull/1794/commits/2b1e37d3b34c5d92797ce7f3b08480035cad98ae

**Comment by omarfsosa** [comment]
This is really cool @juanitorduz ! I've often reused some of those functions but found myself copy-pasting stuff cause I never had the time to write it properly so I really appreciate this.  I think it's looking really good, my only nit would be the naming of the "ell" parameter in some of the functions, but I can see why you did that and tbh I don't have a better suggestion for now.

**Comment by juanitorduz** [comment]
Thanks for the feedback! Exactly! I was copy pasting this a lot as well 😄!

Regarding the variable name: the original name is `L` but I don't find it esthetic as it is upper case and `l` is not encouraged as it is hard to read so I had to go with `ell`. If some of the NumPyro devs have another suggestion I am open 🙂

File: numpyro/contrib/hsgp/approximation.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    length: float,
+    ell: float,
+    m: int,
+    non_centered: bool = True,
+):
+    """
+    Hilbert space Gaussian process approximation using the Matérn kernel.
+
+    The main idea of the approach is to combine the associated spectral density of the
+    Matérn kernel kernel and the spectrum of the Drichlet Laplacian operator to obtain
```
Comment (martinjankowiak) [inline numpyro/contrib/hsgp/approximation.py:115]: Dirichlet

File: numpyro/contrib/hsgp/approximation.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    Stat Comput 30, 419-446 (2020).
+
+    2. Riutort-Mayol, G., Bürkner, PC., Andersen, M.R. et al. Practical Hilbert space
+    approximate Bayesian Gaussian processes for probabilistic programming. Stat Comput 33, 17 (2023).
+
+    :param ArrayImpl x: input data
+    :param float nu: smoothness parameter
+    :param float alpha: amplitude of the squared exponential kernel
+    :param float length: length scale of the squared exponential kernel
+    :param float ell: The length of the interval divided by 2
```
Comment (martinjankowiak) [inline numpyro/contrib/hsgp/approximation.py:130]: ditto on ell?

File: numpyro/contrib/hsgp/laplacian.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+def sqrt_eigenvalues(ell: float, m: int) -> ArrayImpl:
+    """
+    The first `m` square root of eigenvalues of the laplacian operator in `[-ell, ell]`. See Eq. (56) in [1].
+
+    **References:**
+
+    1. Solin, A., Särkkä, S. Hilbert space methods for reduced-rank Gaussian process regression.
+    Stat Comput 30, 419-446 (2020)
+
+    :param float ell: The length of the interval divided by 2.
```
Comment (martinjankowiak) [inline numpyro/contrib/hsgp/laplacian.py:23]: ditto

File: docs/source/contrib.rst [hunk 5]
```diff
@@ -97,3 +97,203 @@ Stochastic Support
     :undoc-members:
     :show-inheritance:
     :member-order: bysource
+
+
+Hilbert Space Gaussian Processes Approximation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This module contains helper functions do use the  Hilbert Space Gaussian Process (HSGP) approximation method
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply docs/source/contrib.rst]: Thanks! Fixed in https://github.com/pyro-ppl/numpyro/pull/1794/commits/d147dd0b97d982e04d34a5a04c54bb284e1b52ef

File: docs/source/contrib.rst [hunk 6]
```diff
@@ -97,3 +97,203 @@ Stochastic Support
...
+Laplacian operator, and :math:`\beta_{j}` are the coefficients of the expansion (see Eq. (8) in [2]).
+
+.. note::
+    Even though the periodic kernel is not stationary, one can still adapt and find a similar approximation formula. 
+    See Appendix B in [2] for more details.
+
+**Example:**
+
+Here is an example of how to use the HSGP approximation method with NumPyro. We will use the squared exponential kernel.
+Other kernels can be used sim
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply docs/source/contrib.rst]: Yes! Fixed in https://github.com/pyro-ppl/numpyro/pull/1794/commits/d147dd0b97d982e04d34a5a04c54bb284e1b52ef

File: docs/source/contrib.rst [hunk 7]
```diff
@@ -97,3 +97,203 @@ Stochastic Support
...
+
+    .. code-block:: python
+
+        >>> from jax import random
+        >>> import jax.numpy as jnp
+
+        >>> import numpyro
+        >>> from numpyro.contrib.hsgp.approximation import hsgp_approximation_squared_exponential
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply docs/source/contrib.rst]: done! :) https://github.com/pyro-ppl/numpyro/pull/1794/commits/d147dd0b97d982e04d34a5a04c54bb284e1b52ef

[the remark this answers is not in the document]
Comment (juanitorduz) [reply docs/source/contrib.rst]: I did a global search and I think we are good now :)

File: numpyro/contrib/hsgp/approximation.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    length: float,
+    ell: float,
+    m: int,
+    non_centered: bool = True,
+) -> ArrayImpl:
+    """
+    Hilbert space Gaussian process approximation using the squared exponential kernel.
+
+    The main idea of the approach is to combine the associated spectral density of the
+    squared exponential kernel and the spectrum of the Drichlet Laplacian operator to
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply numpyro/contrib/hsgp/approximation.py]: +1 https://github.com/pyro-ppl/numpyro/pull/1794/commits/d147dd0b97d982e04d34a5a04c54bb284e1b52ef

File: numpyro/contrib/hsgp/approximation.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+    1. Solin, A., Särkkä, S. Hilbert space methods for reduced-rank Gaussian process regression.
+    Stat Comput 30, 419-446 (2020).
+
+    2. Riutort-Mayol, G., Bürkner, PC., Andersen, M.R. et al. Practical Hilbert space
+    approximate Bayesian Gaussian processes for probabilistic programming. Stat Comput 33, 17 (2023).
+
+    :param ArrayImpl x: input data
+    :param float alpha: amplitude of the squared exponential kernel
+    :param float length: length scale of the squared exponential kernel
+    :param float ell: The length of the interval divided by 2
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply numpyro/contrib/hsgp/approximation.py]: Added better explanation in https://github.com/pyro-ppl/numpyro/pull/1794/commits/d147dd0b97d982e04d34a5a04c54bb284e1b52ef

File: numpyro/contrib/hsgp/approximation.py (hunk 2 above)
Comment (juanitorduz) [reply numpyro/contrib/hsgp/approximation.py]: +1 https://github.com/pyro-ppl/numpyro/pull/1794/commits/d147dd0b97d982e04d34a5a04c54bb284e1b52ef

File: numpyro/contrib/hsgp/laplacian.py (hunk 4 above)
Comment (juanitorduz) [reply numpyro/contrib/hsgp/laplacian.py:23]: Same as above (https://github.com/pyro-ppl/numpyro/pull/1794/commits/d147dd0b97d982e04d34a5a04c54bb284e1b52ef) :)

**Review by fehiepsi — APPROVED** [review]
Thanks @juanitorduz for this great contribution! Thanks @omarfsosa and @martinjankowiak for reviewing! :)

**Pull request merged**